### PR TITLE
Separate OMR x86 linkage properties from linkage class

### PR DESCRIPTION
**Pull request by harpreetbamrah**
Signed-off-by: Harpreet Kaur <hkaur3@unb.ca>

### CODE REVIEW

**Comment by harpreetbamrah** [comment]
This is a WIP PR for x86 linkage refactoring to make properties structure its own thing separate from the linkages. It has been tested on 64-bit and testing for 32-bit is still in progress.

File: compiler/x/codegen/OMRX86LinkageProperties.hpp [hunk 1]
```diff
@@ -0,0 +1,220 @@
+/*******************************************************************************
+ * Copyright (c) 2000, 2019 IBM Corp. and others
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRX86LinkageProperties.hpp:2]: Forward date should be 2020.

File: compiler/x/codegen/OMRX86LinkageProperties.hpp [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef OMRX86LINKAGEPROPERTIES_HPP
+#define OMRX86LINKAGEPROPERTIES_HPP
+
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRX86LinkageProperties.hpp:24]: I would prefer if you would add the `#include` directives necessary for what this file uses.  That way, in can be included anywhere without assuming a context already.  For example (and there are likely others):
```
#include <stdint.h>
#include "codegen/RealRegister.hpp"
```

File: compiler/x/codegen/OMRLinkage.hpp [hunk 3]
```diff
@@ -45,31 +45,7 @@ namespace OMR { typedef OMR::X86::Linkage LinkageConnector; }
...
-
-// register flags
-#define Preserved                   0x01
-#define IntegerReturn               0x02
-#define IntegerArgument             0x04
-#define FloatReturn                 0x08
-#define FloatArgument               0x10
-#define CallerAllocatesBackingStore 0x20
-#define CalleeVolatile              0x40
+#include "x/codegen/OMRX86LinkageProperties.hpp"
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRLinkage.hpp:48]: If the linkage properties header isn't going to be made "extensible" then you don't need the `OMR` prefix.  Perhaps just call it `X86LinkageProperties.hpp`.

Comment (0xdaryl) [reply compiler/x/codegen/OMRLinkage.hpp:48]: The `x/` shouldn't be used unless necessary to disambiguate filenames in different directories.

File: compiler/x/codegen/OMRX86LinkageProperties.hpp [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef OMRX86LINKAGEPROPERTIES_HPP
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRX86LinkageProperties.hpp:22]: When the file is renamed this should probably change to `X86LINKAGEPROPERTIES_HPP`

File: compiler/x/codegen/OMRX86LinkageProperties.hpp (hunk 1 above)
Comment (harpreetbamrah) [reply compiler/x/codegen/OMRX86LinkageProperties.hpp:2]: Thanks for your time @0xdaryl! 

Will have this fixed in next commit.

File: compiler/x/codegen/OMRX86LinkageProperties.hpp (hunk 2 above)
Comment (harpreetbamrah) [reply compiler/x/codegen/OMRX86LinkageProperties.hpp:24]: Will include following:
#include <stdint.h>
#include "codegen/RealRegister.hpp"
#include "codegen/Register.hpp"
#include "codegen/RegisterConstants.hpp"

File: compiler/x/codegen/OMRLinkage.hpp (hunk 3 above)
Comment (harpreetbamrah) [reply compiler/x/codegen/OMRLinkage.hpp:48]: Point noted!

**Comment by harpreetbamrah** [comment]
@0xdaryl, changes tested on Jenkins for sanity.functional test group for x86-32_windows as well as x86-64_linux

**Comment by 0xdaryl** [comment]
It looks like the thrust of this PR is to simply separate the OMR X86 linkage properties from the OMR X86 linkage class, presumably so that it can be consumed in the microJIT.  If that is the case, I think these 4 commits can be squashed down into a single commit whose commit title simply states that action.

**Comment by harpreetbamrah** [comment]
> It looks like the thrust of this PR is to simply separate the OMR X86 linkage properties from the OMR X86 linkage class, presumably so that it can be consumed in the microJIT. If that is the case, I think these 4 commits can be squashed down into a single commit whose commit title simply states that action.

Thanks @0xdaryl for your time!

The main motive behind this PR is the code clean-up as what was inferred from various discussions that the linkage classes are in a bit of a mess cross platform as the APIs vary, but they mostly do the same thing. So, basically these changes intend to make properties structure its own thing separate from the linkages.

And also, architecting the linkage implementation to ultimately re-use the TR linkage properties (in MicroJIT) to keep some commonality between the implementations so as to reduce the need for double maintenance.

Please let me know if further changes are required towards this direction, then I will squash down the commits.

**Comment by 0xdaryl** [comment]
I think you can just squash what you have right now.  It is a step in the journey of refactoring.

**Comment by harpreetbamrah** [comment]
> I think you can just squash what you have right now. It is a step in the journey of refactoring.

All 4 commits squashed into a single commit; good to go now!

**Comment by 0xdaryl** [comment]
Jenkins build win,xlinux,osx

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**